7.improve CIworkflow

ilaflott opened the pull request:
This tries to straighten out the workflow files for githubs CI/CD pipeline. Part of testing these changes is... opening this PR and watching the correct workflows run.

ilaflott commented:
so, almost entirely focused on changing the workflow files here. 

deletes the following, as they are unused/old/redundant
```
.github/workflows/main.yml
.github/workflows/import-test.yml
.github/workflows/docker-image2.yml
.github/workflows/docker-image.yml
```

`.github/workflows/python-conda-env.yml` renamed to `conda-env-create-and-pytest.yml`, expanded upon to include steps originally included in `.github/workflows/create-gfdl-catalog.yml`, which is still present for the moment, but i think we can get rid of.

`meta.yaml` used to specify the `git_url` field with the repo URL. This is a good idea in theory but results in the `conda build .` in the workflow files to only look at code on the main branch- undesirable!!! Thank you @bcc2761 for figuring that out and leaving good commit notes for `fre-cli`. Also now includes a new import test, `catalogbuilder.scripts.gen_intake_gfdl`

`environment.yml` hada a bunch of dependencies that aren't actual dependencies for the `catalogbuilder` package. `environment.yml` now only contains what `catalogbuilder` needs to be pip-installed + run in the environment.

ilaflott commented:
I'm happy with the current tweaks.

Note that all of this is changes regarding the workflow and/or environment files. There are no changes to the code of the package itself.

aradhakrishnanGFDL commented on .github/workflows/conda-env-create-and-pytest.yml:
```diff
@@ -0,0 +1,70 @@
+name: conda-env-create-and-pytest
+
+on: [push]
```
```suggestion
on: [pull_request]
```

aradhakrishnanGFDL commented on .github/workflows/conda-env-create-and-pytest.yml:
```diff
@@ -0,0 +1,70 @@
+name: conda-env-create-and-pytest
```
```suggestion
name: catalog-build-pytests-in-conda-env
```

aradhakrishnanGFDL replied:
the workflow file could also be renamed to catalog-build-pytests-in-conda-env? (or anything that is a bit descriptive as this is an important dynamic catalog building test)

ilaflott replied:
i'm OK with finding a new name, but let's avoid "build" in this context. build is very overloaded here. we have catalogbuilder building catalogs, a pipeline building an environment, a pipeline building a package using conda build... 

ideally, some day, this dynamic catalog creation test exists within a pytest suite

ilaflott replied:
how about `conda-env-create-run-tests`?

aradhakrishnanGFDL commented on meta.yaml:
```diff
@@ -27,7 +27,8 @@ requirements:
     - catalogbuilder
-    - catalogbuilder.scripts
     - catalogbuilder.cats
+    - catalogbuilder.scripts
```
@Ciheim will be curious to see if fre-cli can still take the --config option because I had not moved configs during this restructuring. We can move it if fre-cli requires it, though it shouldn't be required. Please open a separate issue if fre-cli needs catalogbuilder.config

aradhakrishnanGFDL reviewed (changes requested):
Thank you @ilaflott! Minor changes and suggestions made. 
For a later time- This change makes me wonder if a similar, but minimal test to use the catalogbuilder conda package and run the catalog builder might add value as well, upon push. But, I can also see how the analysis-scripts repo might help with this post PR (push) tests.

aradhakrishnanGFDL commented:
@ilaflott not sure if its tied to this PR, but the json and csv files in here https://github.com/NOAA-GFDL/CatalogBuilder/tree/main/catalogbuilder/cats do not show up in the [conda package](https://anaconda.org/NOAA-GFDL/catalogbuilder/files) we install/download from the noaa-gfdl channel. Do you see the same behavior?

Descending into catalogbuilder and looking at cats/ reveals that there is only the init file in there within the published conda pkg. It may have to do something with [this file](https://github.com/NOAA-GFDL/CatalogBuilder/blob/54f92f4d4e38b35ebcd40ad8acdfc4e488c3afd3/MANIFEST.in#L1) that may need a tweak or could be removed given the settings have changed?

ilaflott commented:
thanks for the excellent feedback @aradhakrishnanGFDL, let me take a peek at what you're pointing out RE json files being absent.

as for the conda env building + running test... there is merit to the idea IMO. what makes
me reconsider is the surprising amount of pain github CI/CD introduces to trying to test software within a conda env.

ilaflott commented:
OK, aside from the workflow file name, i've implemented the suggested changes and added another improvement to handling calls to the `conda` environments resources

ilaflott commented:
OK great news! I managed to work the packaging structure out... initially i just wanted to move `tests/` into `catalogbuilder/tests/`, but this caused awkward problems with `sys.path` manipulation for module imports in the tests... and i just couldnt stop going down that rabbit hole.
